### ISSUE TITLE
Read error (array_io_read failed) with ne30 v1scream case

### ISSUE THREAD

**Issue by ndkeen** [E3SM-Project/scream#1505]
Haven't had a chance to look at this closer.

New error with ne30, CPU-only on PM.  This is using 4 nodes, 16 MPI's.  I see the error with OPT build, but not with DEBUG build. 
```
15: array_io_read failed with: /pscratch/sd/n/ndk/wacmy/s17-mar23/externals/ekat/src/ekat/util/ekat_file_utils.hpp:24: FAIL:
15: nread == sz
15: read: nread = 1023 sz = 3000
15: WARNING: SPA Remap File has been set to 'NONE', assuming that SPA data and simulation are on the same grid - skipping horizontal interpolation p3_iso_c::p3_init: One or more table files exists but gave a read error.
 0: WARNING: SPA Remap File has been set to 'NONE', assuming that SPA data and simulation are on the same grid - skipping horizontal interpolation p3_iso_c::p3_init: One or more table files exists but gave a read error.
 0: array_io_read failed with: /pscratch/sd/n/ndk/wacmy/s17-mar23/externals/ekat/src/ekat/util/ekat_file_utils.hpp:24: FAIL:
 0: nread == sz
 0: read: nread = 1023 sz = 3000
15: terminate called after throwing an instance of 'std::logic_error'
15:   what():  /pscratch/sd/n/ndk/wacmy/s17-mar23/components/scream/src/physics/p3/p3_f90.cpp:122: FAIL:
15: info == 0
15: p3_init_c returned info -1
```

```
template<typename T>
void read (T* v, size_t sz, const FILEPtr& fid) {
  size_t nread = fread(v, sizeof(T), sz, fid.get());
  EKAT_REQUIRE_MSG(nread == sz, "read: nread = " << nread << " sz = " << sz);
}
```

could we maybe add the filename that is being read when this error happens?

**Comment by AaronDonahue** [comment]
This is from the table lookups.  If ne4 works we would assume ne30 would too since these are the same tables... right?

**Comment by AaronDonahue** [comment]
sidenote, the `info=-1` referred to here comes from: https://github.com/E3SM-Project/scream/blob/master/components/scream/src/physics/p3/p3_iso_c.f90#L43

There are two places where `info=-1` can be triggered.  It would be easy enough to have `-1` for one of these and `-2` for the other.  It would help us isolate which part of `p3_init_c` failed.

**Comment by ndkeen** [comment]
Yes, there is that print. It is above in the comment, but because of the missing newline in the SPA warning, you have to scroll to the right to see it.

**Comment by AaronDonahue** [comment]
Ahh okay, great.  I should add the `newline` in the SPA warning...

**Comment by ndkeen** [comment]
I ran another test with 8 nodes and 32 MPI's (1 thread) and also see the same error.  Will try other tests.

**Comment by ndkeen** [comment]
I just encountered this error on cori-knl. It runs with ne4, but with ne30 (which requires many nodes), my first job to get beyond OOM hits this error.

**Comment by ndkeen** [comment]
Just noting that I still see this error "sometimes", including with cases that worked before with same PE layout.

**Comment by ndkeen** [comment]
It looks to me like a classic MPI race condition. P3 init is having all MPI ranks read the same files. And note the `array_io_read` function is not only reading from a file, but also opening it.  We probably want one task to read file and then MPI broadcast the data to others.

**Comment by PeterCaldwell** [comment]
Good point Noel. This task is somewhat related to https://github.com/E3SM-Project/scream/issues/1522 (in that our I/O interface isn't perfect yet). Could someone !=Peter C volunteer to issue a PR fixing this?

**Comment by bartgol** [comment]
Shouldn't it be ok to do concurrent read from a file? So long as the file is _already there when ranks attempt to read_. Are the P3 tables files already present when we run? I think they are, but I'm not 100% sure. If they are created on the fly, that might be an issue, and we should have this auto-table-generation mechanism made parallel-safe.

**Comment by bartgol** [comment]
I actually just checked, and the tables are not in the `run/data` folder. As it happens for standalone scream, P3 generates them at init time. But I think the tables init work is done on all ranks, causing potential issues? I will have to double check this.

**Comment by bartgol** [comment]
Yes, it appears all ranks could potentially init the tables. I am not sure how this is not causing corrupted data, since I can definitely see two ranks having the same file open for write and writing at the same time.

**Comment by bartgol** [comment]
> I'm confused. I know some of the p3 data tables are precomputed. They look to be on e3sm's inputdata server rather than in our git repo under data/. Looking at a sim that already ran, these files do seem to be present. _other_ p3 lookups are computed online, but I was pretty sure those online calculations were held in computer memory rather than written to a file (because why would we bother?).

The tables whose files are guaranteed to exist at startup are `p3_lookup_table_1` and `p3_lookup_table_2`. But P3 generates also `mu_r_table_vals.dat`, `revap_table_vals.dat`, `vn_table_vals.dat`, and `vm_table_vals.dat`. I suppose our F90 code writes them, so that we can do the calculation only once for all our unit tests, and then simply load a file. But in a multi-rank framework, this can be dangerous.

We can make it so that the files are written only if a specific flag is passed, which we'll do for unit testing, but won't do for normal runs.

Alternatively, are these tables dependent on runtime parameters? If not, we can put these tables along side the lookup tables on our servers, and simply download them.

**Comment by bartgol** [comment]
There's also the issue of us running multiple copies of the same test, with different ranks. E.g., `p3_standalone_npX` and `p3_standalone_npY` will both try to create tables (if tables are not present already), even if we make each test create tables only on one rank.

However, it seems to me that `p3_init_a` and `p3_init_b` do operations that only depend on lookup tables, and p3 internals, with no input from the outside (e.g., no dependence on number of levels. So maybe we should generate tables for `mu_r`, `revap`, `vn`, and `vm` once and for all, and keep them in the e3sm input data server. Or, if we don't want to put them on the server, since some CMake configurable param _might_ affect calculation, we could generate them once at config time, kind of like rrtmgp generates the all-sky baselines.

Pinging the p3-warriors: @jgfouca @PeterCaldwell @AaronDonahue.

**Comment by jgfouca** [comment]
@bartgol , I haven't looked at this stuff in a while, but I thought I set up dependencies in the testing in CMake so that one test would run before the others to guarantee input tables would be available:

in `scream/src/physics/p3/tests/CMakeLists.txt` :
 
```
# The p3_test_setup executable generates tables used by all p3 tests. This                                                                                                                                                                        
# executable is a test fixture in the sense of CMake, so we mark it with the                                                                                                                                                                      
# FIXTURES_SETUP property, and we make p3_tests and p3_run_and_cmp_* depend on                                                                                                                                                                    
# it via the FIXTURES_REQUIRED property.                                                                                                                                                                                                          
# (See https://cmake.org/cmake/help/latest/prop_test/FIXTURES_REQUIRED.html)                                                                                                                                                                      
CreateUnitTest(p3_test_setup p3_test_setup.cpp "${NEED_LIBS}"
               EXCLUDE_MAIN_CPP PROPERTIES FIXTURES_SETUP p3_tables)
```

**Comment by PeterCaldwell** [comment]
I'm 95% sure that the F90-generated lookup tables in p3_init_b are independent of anything we would ever change. There's a comment in the p3_init_b code saying that "this is very fast so it can be generated at the start of each run"... which I take to mean that it is only being computed online because doing the calculation is as fast as reading in the input data. 

So I think we should just put those files in the inputdata repo and never regenerate them. We should create a standalone code (like create_p3_lookupTable_1.f90-v4.1) which tells us *how* to create those tables though.

**Comment by jgfouca** [comment]
@bartgol that sounds like a good soln, thx!
